### PR TITLE
Revamp the authentication docs

### PR DESCRIPTION
**Pull request by redox**
- Describe in details the different authentication mechanisms & provide simple cURL examples.
 - Demistify the `_sorare_session_id` and `csrf-token` values.
 - Document the OAuth endpoints & flow.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,60 +74,106 @@ mutation SignInMutation($input: signInInput!) {
...
+
+This is exactly the authentication mechanism that happens while browsing the [sorare.com](https://sorare.com) website.
+
+### GraphQL `signIn` mutation & JWT tokens
 
-### With a JWT token
+A better way than cookies to stay authenticated is to rely on [JWT tokens](https://jwt.io/).
 
-To authenticate by JWT you need to request a valid JWT token in the `signIn` mutation :
+We provide JWT tokens within the `signIn` mutation. They can be retrieve using the following mutation:
```
Comment (LeFrenchTouch) [inline README.md:125]: -> _We provide a JWT token as part of the GraphQL `signIn` mutation, and it can be retrieved in the following manner:_

File: README.md [hunk 2]
```diff
@@ -27,60 +74,106 @@ mutation SignInMutation($input: signInInput!) {
...
+        expiredAt
       }
-      ...
     }
   }
+}
 ```
 
-`YOUR_AUD` is a mandatory string parameter that helps identifying your app. Also, tokens generated with an `aud` are scoped to this `aud`. We recommend to use an `aud` reflecting the name of your app - like `myrealappname` - to make it easier to debug & track.
+`<YourAud>` is a mandatory _string_ parameter that helps (us & you) identifying your app. We recommend to use an `aud` reflecting the name of your app - like `myappname` - to make it easier to debug & track.
```
Comment (LeFrenchTouch) [inline README.md:141]: _`<YourAud>` is a required _string_ parameter that helps us, and you, identify your app. We recommend using an `aud` related to your app's name to make it easier to debug and track  (e.g.: `myappname`)._

File: README.md [hunk 3]
```diff
@@ -27,60 +74,106 @@ mutation SignInMutation($input: signInInput!) {
...
+  "variables": { "input": { "email": "<YourEmail>", "password": "<YourHashPassword>" } },
+  "query": "mutation SignInMutation($input: signInInput!) { signIn(input: $input) { currentUser { slug email jwtToken(aud: \"<YourAud>\") { token expiredAt } } errors { message } } }"
+}'
+
+{"data":{"signIn":{"currentUser":{"slug":"<YourSlug>","email":"<YourEmail>","jwtToken":{"token":"<YOUR JWT TOKEN>","expiredAt":"..."}},"errors":[]}}}
 ```
-Authorization: Bearer YOUR_TOKEN
-JWT-AUD: YOUR_AUD
+
+You shall then pass the token with an `Authorization` header alongside a `JWT-AUD` header to all next API requests instead of the `cookie` and the `x-csrf-token` headers:
```
Comment (LeFrenchTouch) [inline README.md:157]: _Once you've retrieved a JWT token, you no longer need to use the `cookie` and `x-csrf-token` headers on your API requests, instead use an `Authorization` header containing the JWT token and a `JWT-AUD` header containing the `aud` string:_

File: README.md [hunk 4]
```diff
@@ -27,60 +74,106 @@ mutation SignInMutation($input: signInInput!) {
...
+    "operationName": "CurrentUserQuery",
+    "query": "query CurrentUserQuery { currentUser { slug email } }"
+}'
+
+{"data":{"currentUser":{"slug":"<YourSlug>","email":"<YourEmail>"}}}
 ```
 
 ## 2FA
 
 For account with 2FA enabled the `signIn` mutation will return an `otpSessionChallenge` instead of the `currentUser`.
```
Comment (LeFrenchTouch) [inline README.md:174]: For account
-> For _accounts_

File: README.md [hunk 5]
```diff
@@ -27,60 +74,106 @@ mutation SignInMutation($input: signInInput!) {
...
 
-You then need to do a second call to the `signIn` mutation providing only the `otpSessionChallenge` and a valid `otpAttempt` :
+You then need to do a second call to the `signIn` mutation providing only the `otpSessionChallenge` and a valid `otpAttempt` as input:
```
Comment (LeFrenchTouch) [inline README.md:176]: as _input_:
-> as _inputs_:

File: README.md [hunk 6]
```diff
@@ -91,13 +184,108 @@ You then need to do a second call to the `signIn` mutation providing only the `o
 }
 ```
 
-## OAuth
+## OAuth / Login with Sorare
+
+If you want to make requests on behalf of other users you can rely on the [OAuth](https://oauth.net/2/) protocol.
+
+Please reach out to us first through our [Help Center](https://help.sorare.com/hc/en-us/requests/new) and ask us to generate you some OAuth credentials.
```
Comment (LeFrenchTouch) [inline README.md:191]: to generate _you some OAuth credentials_.
-> to generate _some OAuth credentials for you_.

File: README.md [hunk 7]
```diff
@@ -91,13 +184,108 @@ You then need to do a second call to the `signIn` mutation providing only the `o
...
+
+They are allowed to:
+
+- Access user's nickname, avatar and wallet address
+- Access details about user's Cards, achievements and favourites
+- Access details about user's auctions, offers and notifications
+
+They are NOT allowed to:
+
+- Access user's email address
```
Comment (LeFrenchTouch) [inline README.md:219]: user's
-> _a_ user's

File: README.md [hunk 8]
```diff
@@ -91,13 +184,108 @@ You then need to do a second call to the `signIn` mutation providing only the `o
...
+- List user's future lineups, compose user's lineups, claim user's rewards
+- Bid on cards, sell user's Cards, make offers, accept offers or initiate a withdrawal
```
Comment (LeFrenchTouch) [inline README.md:221]: user's
-> _a_ user's

File: README.md [hunk 9]
```diff
@@ -91,13 +184,108 @@ You then need to do a second call to the `signIn` mutation providing only the `o
...
+They are allowed to:
+
+- Access user's nickname, avatar and wallet address
+- Access details about user's Cards, achievements and favourites
+- Access details about user's auctions, offers and notifications
+
+They are NOT allowed to:
+
+- Access user's email address
+- List user's future lineups, compose user's lineups, claim user's rewards
```
Comment (LeFrenchTouch) [inline README.md:220]: -> _List a user's future lineups, compose their lineups, claim their rewards_

File: README.md [hunk 10]
```diff
@@ -91,13 +184,108 @@ You then need to do a second call to the `signIn` mutation providing only the `o
...
+
+### Access & Refresh Tokens
+
+First you need to create a "Login with Sorare" link in your app and use the following `href`:
+
+```
+https://sorare.com/oauth/authorize?client_id=<YourUID>&redirect_uri=<YourURLEncodedCallbackURI>&response_type=code&scope=
+```
+
+Once signed in Sorare, the user will be asked to authorize your app and will ultimately be redirected to your `redirect_uri` with a `?code=` query parameter, for instance `https://myapp.com/auth/sorare/callback?code=<YourCode>`.
```
Comment (LeFrenchTouch) [inline README.md:231]: signed in Sorare
-> signed in _to_ Sorare

File: README.md [hunk 11]
```diff
@@ -91,13 +184,108 @@ You then need to do a second call to the `signIn` mutation providing only the `o
...
+```bash
+$ curl -X POST "https://api.sorare.com/oauth/token" \
+-H 'content-type: application/x-www-form-urlencoded' \
+-d 'client_id=<YourOAuthUID>&client_secret=<YourOAuthSecret>&code=<TheRetrievedCode>&grant_type=authorization_code&redirect_uri=<TheSameCallbackURIAsBefore>'
+
+{"access_token":"....","token_type":"Bearer","expires_in":7200,"refresh_token":"...","scope":"public","created_at":1639608238}
+```
 
-If you want to make requests on behalf of other users you can pass an OAuth access token. We'll need to create the OAuth Application for you first. This is done manually for now, on request.
+You can then use the `access_token` like you would use a JWT token:
```
Comment (LeFrenchTouch) [inline README.md:251]: _like_ you would
-> _the same way_ you would

**Review by LeFrenchTouch — APPROVED** [review]
LGTM

**Review by davidbyttow — COMMENTED** [review]
I didn't finish reading but had one major comment.

File: README.md [hunk 12]
```diff
@@ -4,15 +4,62 @@ We are committed to providing an open platform for developers to build upon.
...
-
-Our primary API is specified in the [GraphQL](https://graphql.org/) query language and documented in our [GraphQL playground](https://api.sorare.com/graphql/playground).
+Our API is using the [GraphQL](https://graphql.org/) query language and documented in our [GraphQL playground](https://api.sorare.com/graphql/playground).
 
 ## Authentication
 
-A GQL mutation is available for authentication: `signIn`.
+### Pre-requisites
+
+Before being authenticated with an `Authorization` HTTP header, all requests performed on our GraphQL API need:
```
[the remark this answers is not in the document]
Comment (redox) [reply README.md:13]: I thought about this and was about to remove it but then I also thought about all users currently not using JWT. Any strong opinion @such ? I'm happy removing the "legacy/hacky" cookie-based.

[the remark this answers is not in the document]
Comment (davidbyttow) [reply README.md:13]: Well, I think they should switch to JWT. But also, those already using it, already know how to use it.

File: README.md [hunk 13]
```diff
@@ -27,60 +74,106 @@ mutation SignInMutation($input: signInInput!) {
...
-To use a cookie authentication you need to store the new `_sorare_session_id` and `X-CSRF-TOKEN` and pass it to all next requests.
+{"data":{"currentUser":{"slug":"<YourSlug>","email":"<YourEmail>"}}}
+```
+
+This is exactly the authentication mechanism that happens while browsing the [sorare.com](https://sorare.com) website.
+
+### GraphQL `signIn` mutation & JWT tokens
 
-### With a JWT token
+A better way than cookies to stay authenticated is to rely on [JWT tokens](https://jwt.io/).
```
Comment (davidbyttow) [inline README.md:123]: For short and long-lived authentication, you can request a [JWT](https://jwt.io) token.

File: README.md (hunk 3 above)
Comment (davidbyttow) [reply README.md:157]: (realistically you'll never need a cookie/x-csrf-token with JWT)

File: README.md (hunk 5 above)
Comment (davidbyttow) [reply README.md:176]: or something like:

In this case, you will need to make another call to the `signIn` mutation and provide otpSessionChallenge` value you received and a one-time token from your 2FA device as `otpAttempt`:

File: README.md [hunk 14]
```diff
@@ -91,13 +184,108 @@ You then need to do a second call to the `signIn` mutation providing only the `o
 }
 ```
 
-## OAuth
+## OAuth / Login with Sorare
+
+If you want to make requests on behalf of other users you can rely on the [OAuth](https://oauth.net/2/) protocol.
```
Comment (davidbyttow) [inline README.md:189]: I would reword this a bit. Really this is something like:

"With our OAuth API, user's can sign-in to your service using their Sorare account, which allows you to request data on their behalf."

File: README.md (hunk 6 above)
Comment (davidbyttow) [reply README.md:191]: In order to use our OAuth API, we will need to issue you a Client ID and Secret for your application. You can request one through our [Help Center](https://help.sorare.com/hc/en-us/requests/new) with the following information:
```
1. A unique name for your application
2. One or more callback URLs (e.g., https://yourapp.com/oauth/sorare/callback)
3. A logo for your application in PNG format
```

File: README.md [hunk 15]
```diff
@@ -91,13 +184,108 @@ You then need to do a second call to the `signIn` mutation providing only the `o
...
+We need:
+
+- the name of your App
+- the callback URIs of your OAuth applications
+  - ex: `http://localhost:3000/auth/sorare/callback` (for dev) & `https://myapp.com/auth/sorare/callback` (for prod)
+- the PNG version of your App logo
+
+### OAuth Credentials
+
+Once manually validated, we'll provide you with:
```
Comment (davidbyttow) [inline README.md:202]: Once we validate your application, you will be provided with:
- OAuth Client ID
- OAuth Secret

File: README.md [hunk 16]
```diff
@@ -91,13 +184,108 @@ You then need to do a second call to the `signIn` mutation providing only the `o
...
+
+- OAuth UID (or Client ID)
+- OAuth Secret (keep this secret!)
+
+### OAuth Scopes
+
+No specific scopes are available at the moment. All OAuth apps will be have the same privileges.
```
Comment (davidbyttow) [inline README.md:209]: All OAuth applications are provided with one scope which allows access to the following:
- Basic user information, including their nickname, avatar, and wallet address
- User's cards, achievements and favorites
- User's auctions, offers and notifications

File: README.md [hunk 17]
```diff
@@ -91,13 +184,108 @@ You then need to do a second call to the `signIn` mutation providing only the `o
...
+
+They are allowed to:
+
+- Access user's nickname, avatar and wallet address
+- Access details about user's Cards, achievements and favourites
+- Access details about user's auctions, offers and notifications
+
+They are NOT allowed to:
```
Comment (davidbyttow) [inline README.md:217]: The following are not accessible:
- Email addresses
- Future lineups and rewards
- Claiming rewards
- Bidding, selling, or making offers cards
- Accepting offers or initiating withdrawals

File: README.md [hunk 18]
```diff
@@ -91,13 +184,108 @@ You then need to do a second call to the `signIn` mutation providing only the `o
...
+    "query": "query CurrentUserQuery { currentUser { slug } }"
+}'
+
+{"data":{"currentUser":{"slug":"<ASlug>"}}}
+```
 
 ## Rate limit
 
-The API is rate limited by default. We can provide an API Key on demand that raises the limits. The API key should be passed in an http `APIKEY` header.
+The GraphQL API is rate limited by default. We can provide an extra API Key on demand that raises those limits.
```
Comment (davidbyttow) [inline README.md:267]: I feel like we should rework this and a bit and not even talk about our "unauthenticated" API but instead talk about our API key and mention that one can make calls without the API key but it's drastically limited.

File: README.md [hunk 19]
```diff
@@ -4,15 +4,62 @@ We are committed to providing an open platform for developers to build upon.
 
 While the Cards themselves are stored on the Ethereum blockchain (or in the Starkware rollup) we support an API that provides more detailed information.
```
Comment (davidbyttow) [inline README.md:5]: This is a big first step toward clarification. I think this should be at some point broken down into three distinct sections for authentication:
1. API Key (including unauthenticated) 
2. JWT
3. OAuth

**Review by davidbyttow — CHANGES_REQUESTED** [review]

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (such) [reply README.md:13]: Agreed

[the remark this answers is not in the document]
Comment (such) [reply README.md:13]: But currently to retrieve the JWT token you have to authenticate your request with a cookie right?
We might want to add the feature of retrieving a token in the frontend rather sooner than later

[the remark this answers is not in the document]
Comment (such) [reply README.md:13]: My bad we don't

[the remark this answers is not in the document]
Comment (davidbyttow) [reply README.md:13]: Nope! Just use signIn (unauthenticated) and ask for it.

[the remark this answers is not in the document]
Comment (redox) [reply README.md:13]: Will do!

**Comment by redox** [comment]
Closing this one in favor of https://github.com/sorare/api/pull/63 now that I have the right repository access.
